### PR TITLE
Make quantifiers immutable

### PR DESCRIPTION
**Pull request by SquidDev**
This is just finishing off some work that @slindley started a wee while back:

 - Remove the `quantifier list ref` from `ForAll` types, replacing it with a simple `quantifier list`. This means any corresponding boxing/unboxing is removed, along with the collapsing of them within unification.

 - Quantifiers are now just a pair of variable IDs and kinds, and thus are always rigid. The point is no longer mentioned in the quantifier itself, only in the for aAll body (similar to how mutually recursive types are handled).

As a consequence, all remaining code dealing with flexible quantifiers can be deleted, hopefully simplifying things a little bit.

From my testing, there's only one behavioural change - `(~id) : forall a::Any. forall e::Row. (a) -e-> a` will no longer type check, as `forall a. forall b.` is not considered the same as `forall a, b.`. This feels more correct behaviour to me, as a type application for one has different arities as for the other, but I may be wrong here!

### CODE REVIEW

File: core/types.ml [hunk 1]
```diff
@@ -191,14 +191,12 @@ type tycon_spec = [
...
 
-let unbox_quantifiers = (!)
-let box_quantifiers = ref
 
 (* Generation of fresh type variables *)
 let type_variable_counter = ref 0
-let fresh_raw_variable : unit -> int =
-  function () ->
-    incr type_variable_counter; !type_variable_counter
+let fresh_raw_variable () : int =
```
[the remark this answers is not in the document]
Comment (SquidDev) [reply core/types.ml]: Ahh, no - I think I changed this when trying to trace when variables were introduced, and then forgot to revert it.

**Comment by SquidDev** [comment]
> I'm not sure I understand why there is a difference between `forall a. forall b.` and `forall a,b.` [...] However, if this is actually desired behavior I can live with it.

I had a chat with Sam about this on Friday, and it's definitely not the intended behaviour. The main problem is that type application requires that the number of type variables and type arguments match up exactly. However, `forall a. forall b.` has 1 type variable and `forall a,b.` has 2, so they aren't entirely equivalent.

I don't _think_ this will be too hard to fix - it should be possible to change `Instantiate.apply_type` to handle this, but I haven't looked into it yet.

The alternative would be to change `` `Forall`` to take a single quantifier, which would eliminate this confusion, but I don't know what issues that would introduce.

**Comment by SimonJF** [comment]
I had a look through this PR and everything seems fine to me. I also tried it on my larger project, and everything seems OK (there need to be some changes due to ordered quantification, but nothing specific to this patch).

**Comment by jamescheney** [comment]
Changing `ForAll` to be single-variable-at-a-time would remove the possibility of equivocation between `ForAll([a],ForAll([b],...))` and `ForAll([a,b],...)`  but I think changing `ForAll` and `TAbs`/`TApp` to take single arguments rather than lists would entail fairly large-scale changes.  You're also right that any change that makes the `ForAll` lists misaligned with `TAbs` or `TApp` is not a good idea.

What I was suggesting was instead to bake in the expectation that `ForAll`s can never be immediately nested inside each other.  Before, we didn't notice this distinction because during unification, the variables were rearranged/hoisted anyway.  (But that was causing other problems for generating typed IR too.)

We could, for example, define a smart constructor for `ForAll` that takes a quantifier list and a type, and before building the result checks to see whether the type argument is already a `ForAll`: if so, it merges the quantifier lists, otherwise creates a new `ForAll` node.  Then, it would be impossible to construct a `ForAll([a],ForAll([b],...))`.  This should (hopefully) mean that everything else (including generating appropriate TApp/TAbs code) "just works".

Alternatively, the approach you have explored of trying to align the quantifier lists in unification might be enough, but seems like it might run into similar TApp/TAbs alignment issues.

I'm inclined to say this PR should just go ahead since it's clearly a win on its own, but we should make a note to revisit the quantifier equivocation / normalization issue.

**Pull request merged**